### ISSUE TITLE
HMR is not working

### ISSUE THREAD

**Issue by zhuhaow** [QwikDev/qwik#1901]
### Qwik Version

0.12.0

### Operating System (or Browser)

macOS 

### Which component is affected?

Qwik Rollup / Vite plugin

### Expected Behaviour

Create a website with `pnpm create qwik@latest`, run `pnpm dev`, open a page, change the source code, and expect the page to be updated with HMR.

### Actual Behaviour

The page refreshed.

I checked the ws message, it always got a full reload.

<img width="487" alt="image" src="https://user-images.githubusercontent.com/1287724/198640317-f6b15c78-6192-4bfe-b268-2aaeed523d76.png">

**Comment by FDiskas** [comment]
There are two possible cases when it happens.
> 1. If Vite or a plugin does not handle HMR, a full reload will happen.
> 1. Also if there is a dependency loop, a full reload will happen. To solve this, try removing the loop.

https://vitejs.dev/guide/troubleshooting.html#a-full-reload-happens-instead-of-hmr:~:text=A%20full%20reload%20happens%20instead%20of%20HMR

**Comment by jweb89** [comment]
Just leaving a comment here that this is something I really think needs to be prioritized. It's a very poor developer experience when working on something that requires state in the page and it does a complete refresh when the code is changed

**Comment by rohan-kadkol** [comment]
> Just leaving a comment here that this is something I really think needs to be prioritized. It's a very poor developer experience when working on something that requires state in the page and it does a complete refresh when the code is changed

Losing the state is especially hurtful for us since we make network calls when the page loads. So just adding/removing a small tailwind class triggers a complete refresh that ultimately eats up our API usage quota.

So just wondering, is there any update on this?

**Comment by jweb89** [comment]
@manucorporat Sorry for the ping but I still think this should be a priority. It is probably the single biggest thing as far as DX that would be holding qwik back from bigger adoption especially for more complex sites such as companies who are interested in qwik.

**Comment by jweb89** [comment]
@manucorporat This is not completed. It's still doing a full reload and state is not persevered.
https://stackblitz.com/edit/qwik-starter-ev2u7i?file=src%2Froutes%2Findex.tsx,package.json

**Comment by rohan-kadkol** [comment]
> @manucorporat This is not completed. It's still doing a full reload and state is not persevered. [stackblitz.com/edit/qwik-starter-ev2u7i?file=src%2Froutes%2Findex.tsx](https://stackblitz.com/edit/qwik-starter-ev2u7i?file=src%2Froutes%2Findex.tsx) If you save the file it loses the count

After updating the `qwik` version in the sandbox to `0.20.1` (the latest release), it still doesn't work. ☹️ I think this should be reopened.

**Comment by rohan-kadkol** [comment]
There is a lot of confusion about whether HMR works for Qwik or not. Some repo contributors say it does, while others say they can't get it working. I asked/followed this on [Discord](https://discord.com/channels/842438759945601056/1105863680492580925), [GitHub discussion](https://github.com/BuilderIO/qwik/discussions/3260#discussion-4926889), and this issue, but still am not clear whether HMR works. Because HMR doesn't even work on a brand new Qwik project. 

Can you please clarify if I am doing something wrong? Or does HMR just not work for Qwik?

**Comment by wmertens** [comment]
FWIW I added the example code from Vite to routes/index.tsx and then it will start saying `hmr update` instead of `page reload` when the file changes.

However, this doesn't do anything in the browser.

When you set `debug: true` in qwikVite in the vite plugins file, you can see that the handeHotUpdate hook gets called, but it has no effect.

```
[QWIK PLUGIN: 839] handleHotUpdate() {
```

**Comment by mhevery** [comment]
Not a priority, but if anyone would like to make a PR, I am happy to guide them through the process...

**Comment by ronnyek** [comment]
I wish I knew more about the underlying workings of qwik or HMR to be able to help fix or provide a PR. In the absence of that I would just like to say I'd very much appreciate this functioning (though for me, it wont have significant bearing on whether we adopt qwik or not)

**Comment by wmertens** [comment]
I think the missing piece is getting it to rerender parts that have a changed module.
I suppose that's harder to do because not all rendered HTML has code on the client, so in that case it needs to do SSR.

So maybe this is what's needed:
- inject something browser-side that listens to vite HMR (I think that's already in there)
- be able to replace already-loaded code (probably through a registry that wraps all loaded code)
- re-render when the code is replaced, keeping state the same

On top of that, sometimes a change will require SSR because the changed code only runs on the server.

**Comment by thejackshelton** [comment]
https://vitejs.dev/blog/announcing-vite5-1.html

> On top of that, sometimes a change will require SSR because the changed code only runs on the server.

Our prayers were heard and in vite 5.1 there is now support for HMR w/ SSR.

**Comment by nakasyou** [comment]
Qwik won't support HMR?

**Comment by wmertens** [comment]
> Qwik won't support HMR?

It's not a priority but PRs are certainly welcome

**Comment by wmertens** [comment]
Looks like we'll be getting HMR working via the environment API in vite 6. So, patience :)

**Comment by wmertens** [comment]
Actually I think the client work will be the same if you do it now or in environment, so help welcome!

The idea is that for modules already loaded on the client, they should trigger rerenders of the components that referenced them.

This means: 
- components: just rerender
- handlers, tasks: should just work, next time they run the new module should get used. Maybe for tasks with cleanups, they should be rerun.

For code not on the client, it should somehow figure out which nodes to wake up and rerender. Fallback is to reload the page as before.

**Comment by shairez** [comment]
Re-opening, we'll close this one once the environment API integration is done
